From 93d72f95f87f33d09028ad6035f9d426747caa26 Mon Sep 17 00:00:00 2001
From: Jian-Ming Wang <jian.ming.wang@mitaccomputing.com>
Date: Wed, 30 Apr 2025 08:58:12 +0000
Subject: [PATCH 1/1] Added base board init file for AST2600 and AST2700.
 However, evb_ast2700.c is required to implement based on datasheet.

Upstream-Status: Inappropriate [No upstream maintenance]
---
 board/aspeed/evb_ast2600/evb_ast2600.c | 218 ++++++++++++++++++++++++-
 board/aspeed/evb_ast2700/evb_ast2700.c |  67 ++++++++
 2 files changed, 284 insertions(+), 1 deletion(-)

diff --git a/board/aspeed/evb_ast2600/evb_ast2600.c b/board/aspeed/evb_ast2600/evb_ast2600.c
index e6dc8c7952..0044ba5812 100644
--- a/board/aspeed/evb_ast2600/evb_ast2600.c
+++ b/board/aspeed/evb_ast2600/evb_ast2600.c
@@ -1,5 +1,221 @@
 // SPDX-License-Identifier: GPL-2.0+
 /*
- * Copyright (c) Aspeed Technology Inc.
+ * Copyright (C) ASPEED Technology Inc.
  */
 #include <common.h>
+#include <asm/io.h>
+
+#define SCU_BASE	0x1e6e2000
+#define ESPI_BASE	0x1e6ee000
+#define LPC_BASE	0x1e789000
+#define   LPC_HICR5		(LPC_BASE + 0x80)
+#define   LPC_HICR6		(LPC_BASE + 0x84)
+#define   LPC_SNPWADR	(LPC_BASE + 0x90)
+#define   LPC_HICRB		(LPC_BASE + 0x100)
+#define GPIO_BASE	0x1e780000
+#define AST_WDT1_BASE	0x1e785000
+#define AST_WDT2_BASE	0x1e785040
+
+/* HICR5 Bits */
+#define HICR5_EN_SIOGIO (1 << 31)		/* Enable SIOGIO */
+#define HICR5_EN80HGIO (1 << 30)		/* Enable 80hGIO */
+#define HICR5_SEL80HGIO (0x1f << 24)		/* Select 80hGIO */
+#define SET_SEL80HGIO(x) ((x & 0x1f) << 24)	/* Select 80hGIO Offset */
+#define HICR5_UNKVAL_MASK 0x1FFF0000		/* Bits with unknown values on reset */
+#define HICR5_ENINT_SNP0W (1 << 1)		/* Enable Snooping address 0 */
+#define HICR5_EN_SNP0W (1 << 0)			/* Enable Snooping address 0 */
+
+/* HRCR6 Bits */
+#define HICR6_STR_SNP0W (1 << 0)	/* Interrupt Status Snoop address 0 */
+#define HICR6_STR_SNP1W (1 << 1)	/* Interrupt Status Snoop address 1 */
+
+/* HICRB Bits */
+#define HICRB_EN80HSGIO (1 << 13)	/* Enable 80hSGIO */
+#define DEFAULT_SGPIO_BYTES 10
+#define MAX_SGPIO_BYTES 16
+
+static void __maybe_unused disable_abr_wdt(void)
+{
+	u32 value;
+	value = readl(ASPEED_FMC_WDT2) & 0xfffffffe;
+	puts("Disarm ABR timer...\n");
+	writel(value, ASPEED_FMC_WDT2);
+}
+
+static void __maybe_unused rearm_os_boot_wdt(void)
+{
+	const u32 timout_sec=180;
+	printf("Set watchdog timeout to %d second...\n", timout_sec);
+	writel(timout_sec *  1000000, AST_WDT2_BASE + 0x04);
+	writel(0x4755, AST_WDT2_BASE + 0x08);
+	writel(0x93, AST_WDT2_BASE + 0x0c);
+}
+
+static void __maybe_unused port80h_snoop_init(void)
+{
+	u32 value;
+	/* enable port80h snoop and sgpio */
+	/* set lpc snoop #0 to port 0x80 */
+	value = readl(LPC_SNPWADR) & 0xffff0000;
+	writel(value | 0x80, LPC_SNPWADR);
+
+	/* clear interrupt status */
+	value = readl(LPC_HICR6);
+	value |= HICR6_STR_SNP0W | HICR6_STR_SNP1W;
+	writel(value, LPC_HICR6);
+
+	/* enable lpc snoop #0 and SIOGIO */
+	value = readl(LPC_HICR5) & ~(HICR5_UNKVAL_MASK);
+	value |= HICR5_EN_SIOGIO | HICR5_EN_SNP0W;
+	writel(value, LPC_HICR5);
+
+	/* enable port80h snoop on SGPIO */
+	value = readl(LPC_HICRB) | HICRB_EN80HSGIO;
+	writel(value, LPC_HICRB);
+}
+
+static void __maybe_unused sgpio_init(u8 sgpio_bytes)
+{
+#define SGPIO_CLK_DIV(N)	((N) << 16)
+#define SGPIO_BYTES(N)		((N) << 6)
+#define SGPIO_ENABLE		1
+#define GPIO554			0x554
+#define SCU_414			0x414 /* Multi-function Pin Control #5 */
+#define SCU_414_SGPM_MASK	GENMASK(27, 24)
+
+	u32 value;
+	/* set the sgpio clock to pclk/(2*(5+1)) or ~2 MHz */
+	if  (sgpio_bytes == 0)
+		return;
+
+	if ((sgpio_bytes == 0) || (sgpio_bytes > MAX_SGPIO_BYTES)){
+		printf("Warning: Invalid assignment of sgpio_bytes (%d). It will use default %d bytes instead.\n", sgpio_bytes, DEFAULT_SGPIO_BYTES);
+		sgpio_bytes = DEFAULT_SGPIO_BYTES;
+	}
+	printf("Enable SGPIO with sgpio_bytes: %d\n", sgpio_bytes);
+	value = SGPIO_CLK_DIV(256) | SGPIO_BYTES(sgpio_bytes) | SGPIO_ENABLE;
+	writel(value, GPIO_BASE + GPIO554);
+	writel(readl(SCU_BASE | SCU_414) | SCU_414_SGPM_MASK,
+		SCU_BASE | SCU_414);
+}
+
+
+
+static void __maybe_unused espi_init(void)
+{
+	u32 reg;
+
+	/* skip eSPI init if LPC mode is selected */
+	reg = readl(SCU_BASE + 0x510);
+	if (reg & BIT(6))
+		return;
+
+	/*
+	 * Aspeed STRONGLY NOT recommend to use eSPI early init.
+	 *
+	 * This eSPI early init sequence merely set OOB_FREE. It
+	 * is NOT able to actually handle OOB requests from PCH.
+	 *
+	 * During the power on stage, PCH keep waiting OOB_FREE
+	 * to continue its booting. In general, OOB_FREE is set
+	 * when BMC firmware is ready. That is, the eSPI kernel
+	 * driver is mounted and ready to serve eSPI. However,
+	 * it means that PCH must wait until BMC kernel ready.
+	 *
+	 * For customers that request PCH booting as soon as
+	 * possible. You may use this early init to set OOB_FREE
+	 * to prevent PCH from blocking by OOB_FREE before BMC
+	 * kernel ready.
+	 *
+	 * If you are not sure what you are doing, DO NOT use it.
+	 */
+	reg = readl(ESPI_BASE + 0x000);
+	reg |= 0xef;
+	writel(reg, ESPI_BASE + 0x000);
+
+	writel(0x0, ESPI_BASE + 0x110);
+	writel(0x0, ESPI_BASE + 0x114);
+
+	reg = readl(ESPI_BASE + 0x00c);
+	reg |= 0x80000000;
+	writel(reg, ESPI_BASE + 0x00c);
+
+	writel(0xffffffff, ESPI_BASE + 0x094);
+	writel(0x1, ESPI_BASE + 0x100);
+	writel(0x1, ESPI_BASE + 0x120);
+
+	reg = readl(ESPI_BASE + 0x080);
+	reg |= 0x50;
+	writel(reg, ESPI_BASE + 0x080);
+
+	reg = readl(ESPI_BASE + 0x000);
+	reg |= 0x10;
+	writel(reg, ESPI_BASE + 0x000);
+}
+
+void __maybe_unused acpi_init(void)
+{
+       u32 reg;
+
+       reg = readl(SCU_BASE + 0x510);
+
+       if (reg & BIT(30)) {
+               reg = BIT(5);
+               writel(reg, SCU_BASE + 0x510);
+       } else {
+               reg = readl(SCU_BASE + 0x51c);
+               reg |= BIT(10);
+               writel(reg, SCU_BASE + 0x51c);
+       }
+
+       /* PIN Mux */
+       reg = readl(SCU_BASE + 0x434);
+       reg |= BIT(8) | BIT(9) | BIT(10) | BIT(11) | BIT(12);
+       writel(reg, SCU_BASE + 0x434);
+       /* P0-P1 P2-P3 P4-P5 Passthrough */
+       reg = readl(SCU_BASE + 0x4bc);
+       reg |= BIT(24) | BIT(25) | BIT(26) | BIT(27) | BIT(28) | BIT(29);
+	printf("SCU4BC: %x", reg);
+       writel(reg, SCU_BASE + 0x4bc);
+       /* PBI and PBO */
+       reg = readl(SCU_BASE + 0x418);
+       reg |= BIT(5) | BIT(6);
+       writel(reg, SCU_BASE + 0x418);
+}
+
+int board_early_init_f(void)
+{
+	espi_init();
+	disable_abr_wdt();
+	return 0;
+}
+
+void board_quiesce_devices(void)
+{
+	u32 default_sgpio_bytes = 0;
+	u32 sgpio_bytes;
+
+	printf("Board: %s\n", __FILE__);
+	if (!env_get("openbmconce")) {
+		printf("Enable BMC Kernel boot WDT for robustness.\n");
+		rearm_os_boot_wdt();
+	} else {
+		printf("Found openbmconce in env. Disarm BMC Kernel boot WDT.\n");
+	}
+
+	if (env_get_yesno("port80_snoop") == 1) {
+		printf("Enable Port80 Snoop.\n");
+		port80h_snoop_init();
+		// Port80 snoop is required to eanble SGPIO.
+		default_sgpio_bytes = DEFAULT_SGPIO_BYTES;
+	}
+
+	// When port80_snoop is enabled, it will force to enable sgpio.
+	sgpio_bytes = env_get_ulong("sgpio_bytes", 10, default_sgpio_bytes);
+	if (sgpio_bytes >0)
+		sgpio_init(sgpio_bytes);
+
+	printf("Enable ACPI and Passthrough GPIOs.\n");
+	acpi_init();
+	return;
+}
diff --git a/board/aspeed/evb_ast2700/evb_ast2700.c b/board/aspeed/evb_ast2700/evb_ast2700.c
index b34aa6e168..25530fc34c 100644
--- a/board/aspeed/evb_ast2700/evb_ast2700.c
+++ b/board/aspeed/evb_ast2700/evb_ast2700.c
@@ -2,4 +2,71 @@
 /*
  * Copyright (C) ASPEED Technology Inc.
  */
+#include <common.h>
+#include <env.h>
+#include <asm/arch/platform.h>
+#include <asm/io.h>
+#define DEFAULT_SGPIO_BYTES 10
+#define MAX_SGPIO_BYTES 16
 
+static void __maybe_unused rearm_os_boot_wdt(void)
+{
+	printf("TODO: %s for AST2700.\n", __func__);
+}
+
+static void __maybe_unused port80h_snoop_init(void)
+{
+	printf("TODO: %s for AST2700.\n", __func__);
+}
+
+static void __maybe_unused sgpio_init(u8 sgpio_bytes)
+{
+	printf("TODO: %s for AST2700.\n", __func__);
+}
+
+static void __maybe_unused espi_init(void)
+{
+	printf("TODO: %s for AST2700.\n", __func__);
+}
+
+void __maybe_unused acpi_init(void)
+{
+	printf("TODO: %s for AST2700.\n", __func__);
+}
+
+int board_early_init_f(void)
+{
+	espi_init();
+	disable_abr_wdt();
+	return 0;
+}
+
+void board_quiesce_devices(void)
+{
+	u32 default_sgpio_bytes = 0;
+	u32 sgpio_bytes;
+
+	printf("Board: %s\n", __FILE__);
+	if (!env_get("openbmconce")) {
+		printf("Enable BMC Kernel boot WDT for robustness.\n");
+		rearm_os_boot_wdt();
+	} else {
+		printf("Found openbmconce in env. Disarm BMC Kernel boot WDT.\n");
+	}
+
+	if (env_get_yesno("port80_snoop") == 1) {
+		printf("Enable Port80 Snoop.\n");
+		port80h_snoop_init();
+		// Port80 snoop is required to eanble SGPIO.
+		default_sgpio_bytes = DEFAULT_SGPIO_BYTES;
+	}
+
+	// When port80_snoop is enabled, it will force to enable sgpio.
+	sgpio_bytes = env_get_ulong("sgpio_bytes", 10, default_sgpio_bytes);
+	if (sgpio_bytes >0)
+		sgpio_init(sgpio_bytes);
+
+	printf("Enable ACPI and Passthrough GPIOs.\n");
+	acpi_init();
+	return;
+}
-- 
2.25.1

